registrations/serializers.py

```diff
@@ -66,27 +71,82 @@ class Meta:
     current_waiting_list_count = serializers.SerializerMethodField()
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        if kwargs["context"]["request"].data.get("event", None):
-            event_id = kwargs["context"]["request"].data["event"]
-            event = Event.objects.filter(id=event_id).select_related("publisher")
-            if len(event) == 0:
-                raise DRFPermissionDenied(_("No event with id {event_id}"))
+    publisher = serializers.SerializerMethodField()
+
+    event = JSONLDRelatedField(
+        serializer=EventSerializer,
+        many=False,
+        view_name="event-detail",
+        queryset=Event.objects.all(),
+    )
+
+    created_time = DateTimeField(
+        default_timezone=pytz.UTC, required=False, allow_null=True
+    )
+
+    last_modified_time = DateTimeField(
+        default_timezone=pytz.UTC, required=False, allow_null=True
+    )
+
+    created_by = serializers.StringRelatedField(required=False, allow_null=True)
+
+    last_modified_by = serializers.StringRelatedField(required=False, allow_null=True)
+
+    def __init__(self, instance=None, *args, **kwargs):
+        super().__init__(instance=instance, *args, **kwargs)
+
+        event_dict = kwargs["context"]["request"].data.get("event", None)
+
+        # Check user permissions if creating a new registration.
+        # LinkedEventsSerializer __init__ checks permissions in case of new registration
+        if not instance and isinstance(event_dict, dict) and "@id" in event_dict:
+            event_url = event_dict["@id"]
+            event_id = urllib.parse.unquote(event_url.rstrip("/").split("/")[-1])
             user = kwargs["context"]["user"]
-            if (
-                user.is_admin(event[0].publisher)
-                or kwargs["context"]["request"].method in SAFE_METHODS
-            ):
+
+            try:
+                event = Event.objects.get(pk=event_id)
+                error_message = _(f"User {user} cannot modify event {event}")
+
+                if not user.is_admin(event.publisher):
+                    raise DRFPermissionDenied(error_message)
+
+                if isinstance(user, ApiKeyUser):
+                    # allow updating only if the api key matches instance data source
+                    if event.data_source != user.data_source:
+                        raise DRFPermissionDenied(_(error_message))
+                else:
+                    # allow updating only if event data_source has user_editable_resources set to True
+                    if not event.is_user_editable_resources():
+                        raise DRFPermissionDenied(_(error_message))
+            except Event.DoesNotExist:
                 pass
+
+    def create(self, request, *args, **kwargs):
+        try:
+            instance = super().create(request, *args, **kwargs)
+        except IntegrityError as error:
+            if "duplicate key value violates unique constraint" in str(error):
```

Comment:
This is an interesting way of checking whether a registration exists already or not. Any other way of checking this (rather than relying on an error message)? I mean, I suppose it's valid if there's a very specific reason for doing this this way, but otherwise, seems a bit unconventional.

Reply:
FWIW I don't think there really is a better way to do that. Or do you have some suggestion?

Reply:
Hmm, well, the reason I asked this question was because I'm not actually sure if there's a better way of doing this. 🙂 It might very well be the way to go, and it seemed like a very deliberate choice. I can take a look if there's a better way of doing this, but it might've been considered already, which, in turn, would make my premature search for alternatives a bit redundant.

Reply:
On further consideration, probably fine to leave it as it is; no point in letting this block anything. Relying on a specific string in an error message usually seems a bit flaky to me, and this looks like something that Django could offer tools for... but on the other hand, I can't come up with any suggestions from the top of my head, and this should work just fine.

So feel free to resolve this thread, but if you have, by any chance, looked into this already, please do share before resolving.

Reply:
Unfortunately I don't have better idea at this point. I don't like relaying to string either, but didn't find any better way to do it